File: examples/crud_demo.ts
// FurrDB CRUD Demo (Deno TypeScript)
// Demonstrates basic CRUD operations for a user profile

const HOST = "127.0.0.1";
const PORT = 7070;

// Helper to send a command and receive a response
async function sendCommand(conn: Deno.Conn, cmd: string): Promise<string> {
  const encoder = new TextEncoder();
  const decoder = new TextDecoder();
  await conn.write(encoder.encode(cmd + "\n"));
  let resp = "";
  while (!resp.endsWith("\n")) {
    const buf = new Uint8Array(1024);
    const n = await conn.read(buf);
    if (n === null) break;
    resp += decoder.decode(buf.subarray(0, n));
  }
  return resp.trim();
}

// Main demo logic
const conn = await Deno.connect({ hostname: HOST, port: PORT });
console.log("Connected to FurrDB\n");

// Create (SET user:1 name and email)
console.log("[CREATE] Set user:1 name and email");
console.log(await sendCommand(conn, 'SET user:1:name Alice'));
console.log(await sendCommand(conn, 'SET user:1:email alice@example.com'));

// Read (GET user:1 name and email)
console.log("\n[READ] Get user:1 name and email");
console.log('Name:', await sendCommand(conn, 'GET user:1:name'));
console.log('Email:', await sendCommand(conn, 'GET user:1:email'));

// Update (SET user:1 name)
console.log("\n[UPDATE] Update user:1 name");
console.log(await sendCommand(conn, 'SET user:1:name Alicia'));
console.log('Updated Name:', await sendCommand(conn, 'GET user:1:name'));

// Exists
console.log("\n[EXISTS] Check if user:1:email exists");
console.log('Exists:', await sendCommand(conn, 'EXISTS user:1:email'));

// Delete (DEL user:1 email)
console.log("\n[DELETE] Delete user:1:email");
console.log(await sendCommand(conn, 'DEL user:1:email'));
console.log('Email after delete:', await sendCommand(conn, 'GET user:1:email'));

// List all keys
console.log("\n[KEYS] List all keys");
console.log(await sendCommand(conn, 'KEYS'));

// Exit
await sendCommand(conn, 'EXIT');
conn.close();
console.log("\nConnection closed");

export {};

File: examples/crud_demo.py
import socket

HOST = '127.0.0.1'
PORT = 7070

def send_command(sock, cmd):
    sock.sendall((cmd + '\n').encode())
    resp = b''
    while not resp.endswith(b'\n'):
        chunk = sock.recv(4096)
        if not chunk:
            break
        resp += chunk
    return resp.decode().strip()

with socket.create_connection((HOST, PORT)) as s:
    print('Connected to FurrDB\n')

    # CREATE
    print('[CREATE] Set user:1 name and email')
    print(send_command(s, 'SET user:1:name Alice'))
    print(send_command(s, 'SET user:1:email alice@example.com'))

    # READ
    print('\n[READ] Get user:1 name and email')
    print('Name:', send_command(s, 'GET user:1:name'))
    print('Email:', send_command(s, 'GET user:1:email'))

    # UPDATE
    print('\n[UPDATE] Update user:1 name')
    print(send_command(s, 'SET user:1:name Alicia'))
    print('Updated Name:', send_command(s, 'GET user:1:name'))

    # EXISTS
    print('\n[EXISTS] Check if user:1:email exists')
    print('Exists:', send_command(s, 'EXISTS user:1:email'))

    # DELETE
    print('\n[DELETE] Delete user:1:email')
    print(send_command(s, 'DEL user:1:email'))
    print('Email after delete:', send_command(s, 'GET user:1:email'))

    # KEYS
    print('\n[KEYS] List all keys')
    print(send_command(s, 'KEYS'))

    # EXIT
    send_command(s, 'EXIT')
    print('\nConnection closed')